### DISCUSSION TITLE
Problem about topology volume constraints

### DISCUSSION THREAD

**Discussion by DUT-wansy** [mdolab/dafoam#300] [General]
Hi,

 This is the first time that I use dafoam to solve the discrete sensitivity of topology optimization, and I have a question about the volume constraint. 

 I think it should be different from the volume constraint of shape optimization, because the volume of topology optimization is calculated based on pseudo density, while the shape optimization is calculated based on the actual mesh cells after deformation.

So if I want to add volume constraint to run the pitzDaily case(density-based topology optimization case),  can you give me some suggestion?

**Comment by friedenhe** [comment]
You should be able to use the variableVolSum as the constraint function. Here is an example of how to use it:
https://github.com/mdolab/dafoam/blob/87cef75f382f78f0836f889cbc6da5114a87dee1/tests/runTests_DASimpleFoam.py#L107
Its detailed implementation is here:
https://github.com/mdolab/dafoam/blob/main/src/adjoint/DAObjFunc/DAObjFuncVariableVolSum.C

**Reply by DUT-wansy** [reply]
Thank you for your reply!
After listening to your suggestions, I made the following changes:

1. use the variableVolSum type to define mesh volume in "objFunc"
```
"VOLUME": {
    "part1": {
        "type": "variableVolSum",
        "source": "boxToCell",
        "min": [-50.0, -50.0, -50.0],
        "max": [50.0, 50.0, 50.0],
        "varName": "alphaPorosity",
        "varType": "scalar",
        "component": 0,
        "isSquare": 0,
        "scale": 1.0,
        "addToAdjoint": True,
            },
         },
```

2. Add Volume Contraint in opt task
```
if args.task == "opt":

    optProb = Optimization("opt", objFun=optFuncs.calcObjFuncValues, comm=gcomm)
    DVGeo.addVariablesPyOpt(optProb)
    DVCon.addConstraintsPyOpt(optProb)

    #Add Contraint Volume
    optProb.addCon("VOLUME", lower=0.0 , upper=1e-05 , scale=1)
```
Is there something wrong with my understanding?

**Reply by friedenhe** [reply]
Yes, these settings look good to me. Make sure you add an objective function to optimize.

**Reply by DUT-wansy** [reply]
OK, Thank you for your guidance!
At the same time, I would like to ask if there are any dafoam papers or material recommendations related to topology optimization. All I see on the website are related to aerodynamic optimization.

**Reply by friedenhe** [reply]
No... We don't have topo opt papers. This is a new feature we recently developed, and we haven't fully tested it yet...
